### PR TITLE
ZeRO-1/2: synchronize IPG buffer reuse per buffer

### LINKED ISSUE #8364: ZeRO-1/2 average_tensor serializes backward compute behind the preceding reduction
## Description

ZeRO stage 1/2 with `overlap_comm=true` still serializes the autograd/current stream behind the reduction stream at the start of every `average_tensor()` call:

```python
if self.overlap_comm:
    stream = self.reduction_stream
    if not get_accelerator().resolves_data_dependency():
        for producer_stream in producer_streams:
            stream.wait_stream(producer_stream)
        get_accelerator().current_stream().wait_stream(stream)
```

The producer-to-reduction dependency is required. However, the reverse dependency makes the current stream wait for all previously queued work on `reduction_stream`, including the preceding bucket's collective. When bucket N+1 becomes ready, backward compute is therefore serialized behind bucket N's collective.

This confines overlap to the interval between launching bucket N and preparing bucket N+1. It is especially costly when collective residence is dominated by rank arrival skew rather than data transfer.

## Environment

- Reproduced on DeepSpeed 0.18.7
- Confirmed present by source inspection in 0.19.6 and current `master`
- PyTorch 2.7.0
- 8x H200 on one NVLink node
- ZeRO stage 2
- bf16 communication
- `overlap_comm=true`
- `contiguous_gradients=true`
- `reduce_scatter=true`
- Hugging Face Trainer
- global batch 512, gradient accumulation 2
- 32 samples per rank per micro-batch
- 6,913,191,280 total parameters
- 2,376,274,560 trainable parameters
- 4,536,916,720 frozen parameters
- no CPU or parameter offload

## Profile evidence

The following values are means per rank per micro-batch from eight Kineto traces:

| Metric | Time |
|---|---:|
| Compute kernel union | 489.0 ms |
| NCCL kernel union | 364.0 ms |
| Compute/NCCL overlap | 75.4 ms |
| NCCL pure transfer estimate | 29.9 ms |
| Wait inside collectives | 334.1 ms |
| NCCL outside compute | 288.6 ms |

The traces show the following consistent timeline for non-final gradient buckets:

1. Bucket N launches on `reduction_stream`.
2. Backward compute resumes and prepares bucket N+1.
3. At or immediately before bucket N+1 launches, the compute stream becomes idle while bucket N's NCCL kernel is still resident.
4. Compute resumes only after that preceding collective completes.

Interval attribution assigns NCCL-outside-compute time as follows:

- 212.2 ms (73.5%): the compute stream is stalled on the reduction stream's tail
- 61.6 ms (21.3%): the final bucket is flushed after backward
- 14.8 ms: other collectives

An event replay that keeps NCCL on its own stream, preserves collective order and measured duration, but removes this premature reverse dependency bounds recoverable time at approximately 204 ms per rank per micro-batch. The result is consistent across ranks.

The effect is not explained by transfer bandwidth or SM contention in this profile. NCCL kernels launch with grid `[16, 1, 1]` on 132 SMs. Only about 29.9 ms of 364.0 ms NCCL residence is attributable to transfer. Rank arrival skew dominates: the slowest rank's compute is 652 ms versus a 520 ms rank mean. This makes the reverse dependency particularly visible: one late rank extends the collective, and every rank's compute stream is then held behind that tail at the next bucket enqueue.

## Why this is distinct from existing fixes

- #5606 intentionally added the reverse wait to prevent reuse of an IPG ping-pong buffer before its preceding reduction completed. It also noted a performance gap for short-compute workloads.
- #7805 removed redundant waits after bucket flush and repaired ping-pong index reset, but left the reverse wait inside `average_tensor()`.
- #8061 / #8080 correctly extended the producer-to-reduction dependency to all streams that copied gradients into the IPG bucket. That correctness dependency must remain.

This issue is not proposing removal of producer dependencies. It concerns the location and scope of the reverse dependency that protects buffer reuse.

## Expected behavior

Launching reduction for bucket N should not block unrelated backward compute that produces bucket N+1.

The current/autograd stream should wait for bucket N's reduction only when the same physical IPG ping-pong buffer is about to be reused or overwritten, not whenever `average_tensor()` is called for the next bucket.

## Proposed design direction

Track reduction completion per IPG buffer index:

1. Keep recording all producer streams for each bucket, as implemented by #8080.
2. Make `reduction_stream` wait on those producers before reading the bucket.
3. Record a reduction-complete event for the physical ping-pong buffer after all reduction-stream reads/copies using that buffer have been queued.
4. Before writing into a buffer index again, make the producer/current stream wait on that buffer's completion event.
5. Do not make the current stream wait on the entire reduction stream when merely enqueueing the next bucket.

This should preserve the buffer-reuse safety intended by #5606 and the multi-producer correctness of #8080 while allowing backward compute to overlap the preceding collective.

Special paths that likely need separate handling include oversized parameters, `contiguous_gradients=false`, CPU offload, MoE process groups, and the final bucket flush.

## Validation requirements

A safe change should include:

- regression coverage for #5606's ping-pong buffer reuse race;
- regression coverage for #8061's multi-producer-stream case;
- rapid reuse of both IPG buffer indices;
- gradient, loss, and final-parameter parity against `overlap_comm=false`;
- a CUDA profiler test or reproducible benchmark showing that compute does not stop at the next bucket enqueue while the preceding collective is resident.

I can prepare a draft PR implementing per-buffer completion events if this direction is acceptable to the maintainers.

### PR DESCRIPTION
**Pull request by almightychang**
## Summary

Fixes #8364.

ZeRO-1/2 currently protects IPG ping-pong buffer reuse by making the current/autograd stream wait on the entire reduction stream in `average_tensor()`. That reverse dependency also waits for unrelated work already queued on the reduction stream, serializing the next backward bucket behind the preceding collective.

This change scopes that dependency to the physical IPG buffer being reused:

- preserve the producer-stream to reduction-stream waits added for multi-stream gradient copies;
- record reduction completion per physical IPG buffer after its final reduction-stream consumer;
- wait on that event immediately before the same buffer is overwritten;
- initialize and release event state together with the physical buffers.

Oversized gradients and non-contiguous fallback reductions do not participate in physical IPG buffer reuse. Accelerators that resolve data dependencies implicitly retain their existing path.

## Stream ordering: before and after

Time flows left to right. `P0`/`P1` are producer copies, `R0`/`R1` are reductions, and `E0` is the completion event for physical buffer 0.

### Before: wait on the whole reduction-stream tail

```text
current/autograd  [ P0 ][ backward + P1 ][ wait for reduction_stream ........ ][ next backward ]
                              |                         ^
                              | producer dependency     | broad reverse wait
                              v                         |
reduction_stream         [ wait P0 ][ R0 collective ...][ wait P1 ][ R1 ... ]

Result: preparing or enqueueing the next bucket can stall behind the preceding collective.
```

### After: wait only before reusing the same physical buffer

```text
current/autograd  [ P0 ][ backward + P1 ][ more backward ][ wait E0 ][ overwrite buffer 0 ]
                     |          |                              ^
                     v          v                              | buffer-0 reuse only
reduction_stream  [ wait P0 ][ R0 ... ][ record E0 ][ wait P1 ][ R1 ... ]

buffer index            0                    1                       0 (reused)

Result: producer-to-reduction safety remains, while unrelated backward work overlaps R0.
```

## Correctness invariants

- A reduction never reads a buffer before all producer copies complete.
- A producer never overwrites a physical buffer before that buffer's previous reduction-stream consumers complete.
- Backward compute for a different ping-pong buffer does not wait on the reduction stream's unrelated tail.
- Completion state is isolated by communication dtype and physical buffer index, and survives logical bucket clears until reuse.

## Tests / validation

- `pre-commit run --files deepspeed/runtime/zero/stage_1_and_2.py tests/unit/v1/zero/test_overlap_comm_record_stream.py`
- Local focused suite: 14 passed; CUDA distributed cases skipped on the CPU-only host.
- Modal 2x H200 stress: 10 repeated overlap/non-overlap comparisons across three bucket sizes and delayed reduction-stream work. Losses, full gradients, and final parameters matched.
- Modal 2x H200 related regression selection: overlap on/off, contiguous/non-contiguous gradients, ZeRO stages 1/2/3, multi-bucket overflow, gradient accumulation, unused parameters, and shared backward graphs. The initial run passed 44 cases and exposed one over-constrained new test assertion; after narrowing it to the actual oversized-gradient invariant, the corrected case passed.
- Bucket boundary coverage exercises parameter size at bucket size - 1, equal to bucket size, and bucket size + 1. Bucketed gradients record and wait for reuse; oversized gradients bypass reuse waits.

The correctness suite deliberately delays reduction-stream work to amplify reuse races. I have not repeated the issue's original 8x H200 profiler experiment in this PR; the performance evidence and expected recoverable interval are documented in #8364.

### CODE REVIEW

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 1]
```diff
@@ -114,6 +114,8 @@ def _pad_tensor_by_size(src_tensor, pad_size, dtype, device):
 @dataclass
 class IPGBucket:
     buffer: List[torch.Tensor] = field(default_factory=list)
+    # Completion of the most recent reduction-stream use of each physical buffer.
+    reduction_complete_events: List = field(default_factory=list)
```
[the remark this answers is not in the document]
Comment (almightychang) [reply deepspeed/runtime/zero/stage_1_and_2.py:118]: Seems false positive?

File: tests/unit/v1/zero/test_overlap_comm_record_stream.py [hunk 2]
```diff
@@ -184,3 +218,322 @@ def test_ipg_bucket_clear_resets_copy_streams():
...
+            nonlocal completion_records
+            completion_records += 1
+            original_record(bucket)
+
+        engine.optimizer._record_ipg_buffer_reduction_complete = count_completion_records
+        original_average = engine.optimizer.average_tensor
+
+        def delayed_average(tensor, communication_data_type):
+            with get_accelerator().stream(engine.optimizer.reduction_stream):
+                torch.cuda._sleep(delay_cycles)  #ignore-cuda
```
[the remark this answers is not in the document]
Comment (almightychang) [reply tests/unit/v1/zero/test_overlap_comm_record_stream.py:461]: Fixed in 9dc5ec597. The stress tests now require `get_accelerator().device_name() == "cuda"` before checking or invoking `torch.cuda._sleep`, so BF16-capable NPU/HPU/XPU jobs skip the CUDA-specific delay test. The focused local suite and pre-commit hooks pass.

**Comment by almightychang** [comment]
@codex review

**Review by alanhuangyoo — COMMENTED** [review]
Reviewed as promised on #8364. I read it against `pr8366` rather than running it on multi-GPU. Both places I said I would look at are handled correctly. I have no blocking findings.

## 1. Is the wait on the right event before buffer *reuse*, not only before consumption?

Yes. The ordering is what makes it work, so here it is in full, at the flush in `reduce_independent_p_g_buckets_and_remove_grads`:

1. `reduce_ipg_grads()` runs while `bucket.index` still names the buffer that just filled. Inside it, `_record_ipg_buffer_reduction_complete(bucket)` stores `reduction_stream.record_event()` at `reduction_complete_events[bucket.index]`, and then `bucket.clear()` empties `copy_streams`.
2. Only after that return does `bucket.index = 1 - bucket.index` flip to the other physical buffer.
3. On the next copy, `copy_streams` is empty, so `producer_stream not in bucket.copy_streams` holds and `_wait_for_ipg_buffer_reuse` waits on `reduction_complete_events[new index]`. That is the event from the last reduction that consumed *this* buffer.

So the wait is keyed to the buffer being overwritten, not to whatever was reduced most recently. I was worried that "once per (bucket, stream)" would skip the wait after a flip. It does not, because `clear()` runs inside `reduce_ipg_grads`, before the flip.

Two details I checked because they could have gone wrong:

- **The event is recorded after `copy_grads_in_partition`.** It sits after the consumer loop inside `with get_accelerator().stream(stream)`. It therefore marks the final reduction-stream use of the buffer, not just the all-reduce, and that is the write `grads_in_partition` depends on.
- **Extra-large and non-contiguous fallback reductions do not enter `reduced_ipg_buffer_dtypes`.** Only the `else` branch that reduces `bucket.buffer[bucket.index]` adds the dtype. A reduction that never touched the IPG buffer cannot record an event against it.

## 2. What if a bucket flushes mid-parameter, so one gradient spans two buffers?

This case cannot happen here, so my concern does not apply. A parameter is placed whole:

- if `bucket.elements + param.numel() > reduce_bucket_size`, the bucket is flushed and flipped **first**, and the parameter goes entirely into the fresh buffer at `narrow(0, bucket.elements, numel)` with `elements == 0`
- if `param.numel() > reduce_bucket_size`, it goes to `extra_large_param_to_reduce` and never enters the bucket

No parameter's gradient straddles two buffers, so there is no second event to reconcile.

## Tests

Strong, and aimed at the state machine rather than end-to-end luck. The cases for record-after-consumers, wait-once-per-stream, per-dtype isolation, and `clear()` preserving completion state versus `set_buffers()` resetting it are the ones a scheme like this breaks first.

## One observation outside this PR

This is pre-existing and not something to change here. I have not verified that it matters.

In `reduce_gradients`, the pipeline-parallel branch calls `bucket.add_buffer(torch.empty(reduce_bucket_size))` and then `bucket.index = 0` on every call. `master` has the identical `bucket.buffer.append(...)`, so this PR only keeps `reduction_complete_events` in step with it, which is correct. But if that branch runs each step, it appends a new buffer that `index = 0` never selects. I have not checked whether something else resets the list between steps. I only noticed it because this PR touches that line.

From my side this looks correct.

**Comment by adenzhou1350** [comment]
I have a small follow-up on top of this PR that removes the device-wide synchronization in `allreduce_and_copy()` when the owner-reduce path has no retained non-contiguous gradients to clear. The synchronization remains whenever `previous_reduced_grads` is non-empty.

The rebased follow-up commit is https://github.com/adenzhou1350/DeepSpeed/commit/79edce8f905f7f832e2cb2bb9185482909318d78.

Validation on top of the current PR head:

- 15 current-tree focused CPU tests passed;
- 8 focused distributed tests previously passed on two RTX 5090 GPUs, including delayed owner-reduce comparisons;
- Qwen3-4B-Base, sequence 2048, BF16, ZeRO-2 DP4, owner-reduce route, weighted 50M/26M bucket portfolio;
- exact candidate/reference equality for all step losses, global norms, and 1,194 final-parameter samples;
- four order-balanced block speedups: `1.03831x`, `1.04268x`, `1.03616x`, `1.05783x`;
- median `1.04050x`, bootstrap median 95% interval `[1.03616, 1.05783]`.

The result is limited to the owner-reduce route on this four-GPU PCIe workload. I do not claim a benefit for the default multi-rank route or other hardware/topologies. If you prefer, I can keep it as a follow-up PR after #8366 merges; otherwise the commit is ready to cherry-pick here.
